In [ ]:
/content/drive/MyDrive/Colab Notebooks/car_evaluation.csv

In [12]:
from sklearn.datasets import load_breast_cancer
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
import seaborn as sns

In [13]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
import tensorflow as tf
from tensorflow import keras

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/car_evaluation.csv", index_col=None)
df

vhigh vhigh.1      2   2.1  small   low  unacc
0     vhigh   vhigh      2     2  small   med  unacc
1     vhigh   vhigh      2     2  small  high  unacc
2     vhigh   vhigh      2     2    med   low  unacc
3     vhigh   vhigh      2     2    med   med  unacc
4     vhigh   vhigh      2     2    med  high  unacc
...     ...     ...    ...   ...    ...   ...    ...
1722    low     low  5more  more    med   med   good
1723    low     low  5more  more    med  high  vgood
1724    low     low  5more  more    big   low  unacc
1725    low     low  5more  more    big   med   good
1726    low     low  5more  more    big  high  vgood

[1727 rows x 7 columns]

컬럼명이 없기 때문에 index_col=0이 아닌 index_col=None으로 해야함

In [14]:
df.columns = ['price', 'maint', 'doors', 'persons', 'lug_capacity', 'safety','output']

In [15]:
df.columns

Index(['price', 'maint', 'doors', 'persons', 'lug_capacity', 'safety',
       'output'],
      dtype='object')

In [16]:
df

price  maint  doors persons lug_capacity safety output
0     vhigh  vhigh      2       2        small    med  unacc
1     vhigh  vhigh      2       2        small   high  unacc
2     vhigh  vhigh      2       2          med    low  unacc
3     vhigh  vhigh      2       2          med    med  unacc
4     vhigh  vhigh      2       2          med   high  unacc
...     ...    ...    ...     ...          ...    ...    ...
1722    low    low  5more    more          med    med   good
1723    low    low  5more    more          med   high  vgood
1724    low    low  5more    more          big    low  unacc
1725    low    low  5more    more          big    med   good
1726    low    low  5more    more          big   high  vgood

[1727 rows x 7 columns]

In [17]:
df.isnull().sum()

price           0
maint           0
doors           0
persons         0
lug_capacity    0
safety          0
output          0
dtype: int64

결측치 확인 -> 결과 : 결측치 없음 !

In [20]:
# 범주형 데이터를 숫자로 변환(엔코딩)
label_encoders = {}
for col in df.columns:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])
    label_encoders[col] = le

In [21]:
df

price  maint  doors  persons  lug_capacity  safety  output
0         3      3      0        0             2       2       2
1         3      3      0        0             2       0       2
2         3      3      0        0             1       1       2
3         3      3      0        0             1       2       2
4         3      3      0        0             1       0       2
...     ...    ...    ...      ...           ...     ...     ...
1722      1      1      3        2             1       2       1
1723      1      1      3        2             1       0       3
1724      1      1      3        2             0       1       2
1725      1      1      3        2             0       2       1
1726      1      1      3        2             0       0       3

[1727 rows x 7 columns]

In [23]:
df['output'].value_counts()

output
2    1209
0     384
1      69
3      65
Name: count, dtype: int64

In [25]:
# 데이터와 타겟 분리
X = df.drop('output', axis=1).values
y = df['output'].values

df에서 output에 대한것을 드랍해라 그래서 그것을 y로

In [26]:
X

array([[3, 3, 0, 0, 2, 2],
       [3, 3, 0, 0, 2, 0],
       [3, 3, 0, 0, 1, 1],
       ...,
       [1, 1, 3, 2, 0, 1],
       [1, 1, 3, 2, 0, 2],
       [1, 1, 3, 2, 0, 0]])

numpy와 dataFrame의 차이점
: dataframe은 테이블 형태, numpy는 배열
dataframe은 모두 numpy로 바꿔야함
분석할 때는 dataframe으로 해야 편하고 최종적으로 분류할 때 numpy로 한다.

In [27]:
y

array([2, 2, 2, ..., 2, 1, 3])

X에는 데이터의 특성들이 들어가고 y에는 정답값들 !!

In [28]:
# 스케일러 -> X의 값을 0과 1 사이로 매핑 (스케일을 줄여주는 것) Y값은 스케일러 하면 안됨
scaler = StandardScaler()
X = scaler.fit_transform(X)

# 테스트를 위한 feature, 테스트를 위한 정답 값
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [29]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape,

((1381, 6), (346, 6), (1381,), (346,))

shape가 의미 하는 것 6개 짜리 컬럼이 1381개 있다는 것
6개 짜리 컬럼이 346개 있다는 것

In [30]:
print ("==================================")
print (" ACCURACY OF OTHER ML CLASSIFIERS" )
print ("==================================")

 ACCURACY OF OTHER ML CLASSIFIERS


In [31]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [32]:
from sklearn.svm import SVC

clf_svm = SVC(random_state=0)
clf_svm.fit(X_train, y_train)

pred_svm = clf_svm.predict(X_test)

print("\n--- SVM Classifier ---")
print(accuracy_score(y_test, pred_svm))
print(confusion_matrix(y_test, pred_svm))


--- SVM Classifier ---
0.8786127167630058
[[ 62   1  14   0]
 [  9   4   0   1]
 [ 17   0 222   0]
 [  0   0   0  16]]


In [33]:
# Build a logistic regression classifier and predict

clf_lr = LogisticRegression(random_state=0)
clf_lr.fit(X_train, y_train)

pred_lr = clf_lr.predict(X_test)

print ("\n--- Logistic Regression Classifier ---")
print (accuracy_score(y_test, pred_lr))
print (confusion_matrix(y_test, pred_lr))


--- Logistic Regression Classifier ---
0.6907514450867052
[[ 18   0  55   4]
 [  2   0  12   0]
 [ 16   0 221   2]
 [  4   0  12   0]]


In [34]:
clf_nn = MLPClassifier(random_state=0)
clf_nn.fit(X_train, y_train)

pred_nn = clf_nn.predict(X_test)

print ("\n--- Neural Network Classifier ---")
print (accuracy_score(y_test, pred_nn))
print (confusion_matrix(y_test, pred_nn))


--- Neural Network Classifier ---
0.9653179190751445
[[ 74   2   1   0]
 [  1   9   0   4]
 [  2   0 237   0]
 [  2   0   0  14]]


/usr/local/lib/python3.11/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [35]:
# Build a decision tree classifier and predict

clf_dt = DecisionTreeClassifier(random_state=0)
clf_dt.fit(X_train, y_train)

pred_dt = clf_dt.predict(X_test)

print ("\n--- Decision Tree Classifier ---")
print (accuracy_score(y_test, pred_dt))
print (confusion_matrix(y_test, pred_dt))


--- Decision Tree Classifier ---
0.976878612716763
[[ 74   0   3   0]
 [  0  12   0   2]
 [  0   0 239   0]
 [  3   0   0  13]]


In [36]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

print ("\n--- Radom Forest ---")
rf_clf = RandomForestClassifier(random_state=0)
rf_clf.fit(X_train, y_train)
pred = rf_clf.predict(X_test)
print(accuracy_score(y_test,pred))
print (confusion_matrix(y_test, pred))


--- Radom Forest ---
0.9739884393063584
[[ 72   1   4   0]
 [  1  10   0   3]
 [  0   0 239   0]
 [  0   0   0  16]]
